File: tmpry/download_nvd.py
import asyncio
import os
from aiohttp import ClientSession, TCPConnector
import datetime

class NVD_Source:
    def __init__(self):
        self.cache_dir_path = "/home/escan/181023/data_source"
        self.cache_dir_name = "nvd"
        self.cachedir = os.path.join(self.cache_dir_path, self.cache_dir_name)
        self.session = None

    async def download_cve_data(self):
        if not self.session:
            connector = TCPConnector(limit_per_host=19)
            self.session = ClientSession(connector=connector)

        # Create the cache directory if it doesn't exist
        os.makedirs(self.cachedir, exist_ok=True)

        current_year = datetime.datetime.now().year
        base_url = "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-"

        # Create a list to store downloaded file paths
        downloaded_files = []

        for year in range(2002, current_year + 1):
            url_to_cve_data = f"{base_url}{year}.json.gz"
            filename = url_to_cve_data.split("/")[-1]
            filepath = os.path.join(self.cachedir, filename)

            # Check if the file already exists in the cache directory
            if not os.path.exists(filepath):
                async with self.session.get(url_to_cve_data) as response:
                    response.raise_for_status()

                    # Open the file and write the content
                    with open(filepath, "wb") as file:
                        while True:
                            chunk = await response.content.read(1024)
                            if not chunk:
                                break
                            file.write(chunk)
            
            # Append the downloaded file path to the list
            downloaded_files.append(filepath)

        # Close the session when done
        await self.session.close()
        self.session = None

        # Sort the downloaded files based on their filenames
        downloaded_files.sort()

if __name__ == "__main__":
    downloader = NVD_Source()
    asyncio.run(downloader.download_cve_data())

File: tmpry/download_nvd.cpp
#include <iostream>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <ctime>
#include <cstring>
#include <algorithm>
#include <vector>
#include <curl/curl.h>

// File download callback function
static size_t WriteToFileCallback(void* contents, size_t size, size_t nmemb, void* userData) {
    std::ofstream* file = static_cast<std::ofstream*>(userData);
    file->write(static_cast<const char*>(contents), size * nmemb);
    return size * nmemb;
}

class NVD_Source {
public:
    NVD_Source() {
        cacheDirPath = "/home/escan/181023/data_source";
        cacheDirName = "nvd";
        cacheDir = cacheDirPath + "/" + cacheDirName;
        session = nullptr;
    }

    ~NVD_Source() {
        if (session) {
            curl_easy_cleanup(session);
        }
    }

    void downloadCVEData() {
        if (!session) {
            session = curl_easy_init();
            curl_easy_setopt(session, CURLOPT_TCP_KEEPALIVE, 1L);
        }

        // Create the cache directory if it doesn't exist
        std::string cmd = "mkdir -p " + cacheDir;
        system(cmd.c_str());

        int currentYear = getCurrentYear();
        std::string base_url = "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-";

        // Create a list to store downloaded file paths
        std::vector<std::string> downloadedFiles;

        for (int year = 2002; year <= currentYear; ++year) {
            std::string urlToCVEData = base_url + std::to_string(year) + ".json.gz";
            std::string filename = urlToCVEData.substr(urlToCVEData.rfind('/') + 1);
            std::string filepath = cacheDir + "/" + filename;

            // Check if the file already exists in the cache directory
            std::ifstream file(filepath);
            if (!file) {
                std::ofstream outfile(filepath, std::ofstream::binary);
                downloadFile(urlToCVEData, &outfile);
                outfile.close();
            }

            // Append the downloaded file path to the list
            downloadedFiles.push_back(filepath);
        }

        // Sort the downloaded files based on their filenames
        std::sort(downloadedFiles.begin(), downloadedFiles.end());

        // Print the downloaded files
        //std::cout << "Downloaded Files:" << std::endl;
        //for (const std::string& file : downloadedFiles) {
            //std::cout << file << std::endl;
        //}

	std::cout << "NVD CVE Data Downloaded and Stored. " << std::endl;
        //for (const std::string& file : downloadedFiles) {
            //std::cout << file << std::endl;
        //}

    }

private:
    std::string cacheDirPath;
    std::string cacheDirName;
    std::string cacheDir;
    CURL* session;

    int getCurrentYear() {
        std::time_t currentTime = std::time(nullptr);
        std::tm* timeinfo = std::localtime(&currentTime); // Use currentTime
        return timeinfo->tm_year + 1900;
    }

    void downloadFile(const std::string& url, std::ofstream* file) {
        CURLcode res;
        curl_easy_setopt(session, CURLOPT_URL, url.c_str());
        curl_easy_setopt(session, CURLOPT_WRITEFUNCTION, WriteToFileCallback);
        curl_easy_setopt(session, CURLOPT_WRITEDATA, file);
        res = curl_easy_perform(session);
        if (res != CURLE_OK) {
            fprintf(stderr, "Failed to download file: %s", curl_easy_strerror(res));
        }
    }
};

int main() {
    curl_global_init(CURL_GLOBAL_DEFAULT);
    NVD_Source downloader;
    downloader.downloadCVEData();
    curl_global_cleanup();
    return 0;
}

